File: 279_Perfect_Squares.cpp
#include <iostream>
#include <cmath>

using namespace std;

class Solution {
public:
    int numSquares(int n) {
        unsigned int inf = INT32_MAX;
        int *dp = new int[n + 1];
        for (int j = 1; j <= n; ++j) {
            dp[j] = j;
        }
        for (int i = 1; i <= ceil(sqrt(n)); ++i) {
            dp[i * i] = 1;
        }
        for (int k = 1; k <= n; ++k) {
            for (int i = (int) sqrt(k); i * i * dp[k] > k; --i) {
                dp[k] = min(dp[k], 1 + dp[k - i*i]);
            }
        }
        return dp[n];
    }
};

int main() {
    cout << "res " << Solution().numSquares(9975) << endl;
    return 0;
}

File: 279_Perfect_Squares.py
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [float('inf') for i in xrange(n + 1)]
        t = 1
        while t != int(n ** 0.5) + 1:
            dp[t * t] = 1
            t += 1
        for i in xrange(1, n + 1):
            for j in xrange(1, int(i**0.5)):
                dp[i] = min(dp[i], 1 + dp[i - j*j])
            pass
        return dp[n]


print Solution().numSquares(9975)
